x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java

```diff
@@ -0,0 +1,265 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.mock.orig.Mockito.verify;
+import static org.elasticsearch.mock.orig.Mockito.when;
+import static org.elasticsearch.xpack.core.snapshotlifecycle.history.SnapshotLifecycleTemplateRegistry.SLM_POLICY_NAME;
+import static org.elasticsearch.xpack.core.snapshotlifecycle.history.SnapshotLifecycleTemplateRegistry.SLM_TEMPLATE_NAME;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verifyZeroInteractions;
+
+public class SnapshotLifecycleTemplateRegistryTests extends ESTestCase {
+    private SnapshotLifecycleTemplateRegistry registry;
+    private NamedXContentRegistry xContentRegistry;
+    private ClusterService clusterService;
+    private ThreadPool threadPool;
+    private Client client;
+
+    @Before
+    public void createRegistryAndClient() {
+        threadPool = mock(ThreadPool.class);
+        when(threadPool.getThreadContext()).thenReturn(new ThreadContext(Settings.EMPTY));
+        when(threadPool.generic()).thenReturn(EsExecutors.newDirectExecutorService());
+
+        client = mock(Client.class);
+        when(client.threadPool()).thenReturn(threadPool);
+        AdminClient adminClient = mock(AdminClient.class);
+        IndicesAdminClient indicesAdminClient = mock(IndicesAdminClient.class);
+        when(adminClient.indices()).thenReturn(indicesAdminClient);
+        when(client.admin()).thenReturn(adminClient);
+        doAnswer(invocationOnMock -> {
+            ActionListener<AcknowledgedResponse> listener =
+                (ActionListener<AcknowledgedResponse>) invocationOnMock.getArguments()[1];
+            listener.onResponse(new TestPutIndexTemplateResponse(true));
+            return null;
+        }).when(indicesAdminClient).putTemplate(any(PutIndexTemplateRequest.class), any(ActionListener.class));
+
+        clusterService = mock(ClusterService.class);
+        List<NamedXContentRegistry.Entry> entries = new ArrayList<>(ClusterModule.getNamedXWriteables());
+        entries.addAll(Arrays.asList(
+            new NamedXContentRegistry.Entry(LifecycleType.class, new ParseField(TimeseriesLifecycleType.TYPE),
+                (p) -> TimeseriesLifecycleType.INSTANCE),
+            new NamedXContentRegistry.Entry(LifecycleAction.class, new ParseField(DeleteAction.NAME), DeleteAction::parse)));
+        xContentRegistry = new NamedXContentRegistry(entries);
+        registry = new SnapshotLifecycleTemplateRegistry(Settings.EMPTY, clusterService, threadPool, client, xContentRegistry);
+    }
+
+    public void testThatNonExistingTemplatesAreAddedImmediately() {
```

Comment:
It looks like most of these are copied from the watcher template registry tests, I'm not sure if it makes sense to have exhaustive tests for each (probably a different discussion).

I also see why there's so much mocking (copied over from the other test implementation). If we keep these mocks that's okay (though I greatly prefer we move away from them), but I understand why they're here. It's up to you whether to leave them or change them

Reply:
I'm not opposed to changing the mocks - I did a little other code cleanup when bringing these tests over too so I don't see why I shouldn't do that as well.

My position is that if we're going to use a superclass to share code, then the concrete subclasses should all test the behaviors they rely on (which is why I didn't bring over *all* the tests - the Watcher version depends on some behavior that this one doesn't need).  That way, if e.g. Watcher is radically refactored or removed, etc. and the Watcher template registry is removed then these tests will still verify the behavior that SLM relies on.

I would like to factor out some of the shared logic so that it's less copy/paste-y, but I didn't want to start making that kind of change to the Watcher tests in this feature branch. Ideally I would have refactored the tests as well in the PR that factored out the shared superclass, but I didn't want to delay this PR for another round of changes to Watcher code.

Reply:
I've changed these tests to remove all the mocks, instead using ClusterServiceUtils, TestThreadPool, VerifyingClient, etc. as recommended.